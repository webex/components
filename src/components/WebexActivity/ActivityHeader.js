import React from 'react';
import PropTypes from 'prop-types';
import {format, isToday, isSameWeek, isYesterday} from 'date-fns';

import WebexAvatar from '../WebexAvatar/WebexAvatar';
import {usePerson} from '../hooks';

import './ActivityHeader.scss';

/**
 * Returns a formatted timestamp based on the given date's offset from the current time.
 *
 * @param {Date} timestamp Date instance to format
 * @returns {string} formattedDate
 */
export function formatMessageDate(timestamp) {
  let formattedDate;

  if (isToday(timestamp)) {
    // 12:00 PM
    formattedDate = format(timestamp, 'p');
  } else if (isYesterday(timestamp)) {
    // Yesterday 12:00 PM
    formattedDate = `Yesterday ${format(timestamp, 'p')}`;
  } else if (isSameWeek(timestamp, new Date())) {
    // Monday 12:00 PM
    formattedDate = format(timestamp, 'iiii p');
  } else {
    // 1/1/2020 12:00 PM
    formattedDate = format(timestamp, 'M/d/yy p');
  }

  return formattedDate;
}

export default function ActivityHeader({personID, created}) {
  const {displayName} = usePerson(personID);

  return (
    <div className="activity-header">
      <WebexAvatar personID={personID} />
      <div className="activity-author">
        <span>{displayName}</span>
        <span className="activity-timestamp">{formatMessageDate(new Date(created))}</span>
      </div>
    </div>
  );
}

ActivityHeader.propTypes = {
  personID: PropTypes.string.isRequired,
  created: PropTypes.string.isRequired,
};
