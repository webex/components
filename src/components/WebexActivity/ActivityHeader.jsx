import React from 'react';
import PropTypes from 'prop-types';
import {
  format,
  isToday,
  isSameWeek,
  isYesterday,
} from 'date-fns';

import WebexAvatar from '../WebexAvatar/WebexAvatar';
import {WEBEX_COMPONENTS_CLASS_PREFIX} from '../../constants';
import {usePerson} from '../hooks';

/**
 * Returns a formatted timestamp based on the given date's offset from the current time.
 *
 * @param {Date} timestamp Date instance to format
 * @returns {string} formattedDate
 */
export function formatMessageDate(timestamp) {
  let formattedDate;

  if (isToday(timestamp)) {
    // 12:00 PM
    formattedDate = format(timestamp, 'p');
  } else if (isYesterday(timestamp)) {
    // Yesterday, 12:00 PM
    formattedDate = `Yesterday, ${format(timestamp, 'p')}`;
  } else if (isSameWeek(timestamp, new Date())) {
    // Monday, 12:00 PM
    formattedDate = format(timestamp, 'iiii, p');
  } else {
    // 1/1/2020, 12:00 PM
    formattedDate = format(timestamp, 'P, p');
  }

  return formattedDate;
}

/**
 * ActivityHeader component displays the header content of an activity.
 * Header includes the avatar of the activity author as well as the
 * activity timestamp.
 *
 * @param {object} props
 * @returns {object} JSX of the component
 */
export default function ActivityHeader({personID, timestamp}) {
  const {displayName} = usePerson(personID);

  return (
    <div className={`${WEBEX_COMPONENTS_CLASS_PREFIX}-activity-header`}>
      <WebexAvatar personID={personID} />
      <div className="activity-author">
        <span>{displayName}</span>
        <span className="activity-timestamp">{formatMessageDate(new Date(timestamp))}</span>
      </div>
    </div>
  );
}

ActivityHeader.propTypes = {
  personID: PropTypes.string.isRequired,
  timestamp: PropTypes.string.isRequired,
};
