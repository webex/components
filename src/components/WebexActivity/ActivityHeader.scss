$content-left-margin: 3.5rem; //avatar width + margin

.activity-header {
  display: flex;
  flex-direction: column;
  position: relative;

  .activity-author {
    @include noselect();
    display: flex;
    min-width: 18.75rem !important;
    margin-left: $content-left-margin;
    font-size: 0.875rem;
    line-height: 1.375rem;
    text-align: left;
    color: $md-gray-70;
    word-wrap: break-word;
  }

  .md-avatar {
    position: absolute;
    margin: 0 0.5rem;
  }

  .activity-timestamp {
    margin-left: 0.5rem;
  }
}
