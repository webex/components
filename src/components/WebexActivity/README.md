# Webex Activity Component

Webex activity component displays the activity content.

<p align="center">
  <img src="./WebexActivity.png" alt="Default Webex Activity" />
</p>

## Preview

To see all the different possible states of the Webex Activity component, you can run our Storybook:

```shell
  npm start
```

## Embed

1. Create a component adapter from which the data will be retrieved (See [adapters](../../adapters)). For instance:

    ```js
    const jsonAdapter = new WebexJSONAdapter(jsonData);
    ```

2. Create a component instance by passing the activity ID as a string
enclose it within [a data provider](../WebexDataProvider/WebexDataProvider.js)
that takes the [component data adapter](../../adapters/WebexJSONAdapter.js) that we created previously

    ```js
    <WebexDataProvider adapter={jsonAdapter}>
      <WebexActivities activityID="activityID" />
    </WebexDataProvider>
    ```

The component knows how to manage its data. If anything changes in the data source that the adapter manages, the component will also update on its own.
