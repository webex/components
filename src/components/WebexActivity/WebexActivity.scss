$content-left-margin: 3.5rem; //avatar width + margin
.activity {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  width: 100%;
  margin: 0.5rem 0;

  .activity-header {
    display: flex;
    position: relative;

    .activity-author {
      @include noselect();
      display: flex;
      min-width: 18.75rem !important;
      margin-left: $content-left-margin;
      font-size: 0.875rem;
      line-height: 1.375rem;
      text-align: left;
      color: $md-gray-70;
      word-wrap: break-word;
    }

    .md-avatar {
      position: absolute;
      margin: 0 0.5rem;
    }

    .activity-timestamp {
      margin-left: 0.5rem;
    }
  }

  .activity-content {
    display: flex;
    align-items: flex-start;
    box-sizing: border-box;
    min-width: 15.625rem !important;
    margin-left: $content-left-margin;
    font-size: 1rem;
    color: $black;
    word-wrap: break-word;
    white-space: pre-wrap;
  }
}
