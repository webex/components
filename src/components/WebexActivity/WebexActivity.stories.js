import React from 'react';
import {storiesOf} from '@storybook/react';
import {addDays, getDay, subDays} from 'date-fns';

import jsonData from '../../data';
import {WebexJSONAdapter} from '../../adapters';
import {WebexActivity, WebexDataProvider} from '..';

// Setup for the stories
const stories = storiesOf('Webex Activity', module);
const adapter = new WebexJSONAdapter(jsonData);

// Stories
stories.add('default', () => (
  <WebexDataProvider adapter={adapter}>
    <WebexActivity activityID="default" />
  </WebexDataProvider>
));

stories.add('no header', () => (
  <WebexDataProvider adapter={adapter}>
    <WebexActivity activityID="no-header" />
  </WebexDataProvider>
));

stories.add('multi-line text', () => (
  <WebexDataProvider adapter={adapter}>
    <WebexActivity activityID="multi-line" />
  </WebexDataProvider>
));

stories.add('long text', () => (
  <WebexDataProvider adapter={adapter}>
    <WebexActivity activityID="long" />
  </WebexDataProvider>
));

stories.add('created today', () => {
  const today = new Date().toString();

  jsonData.activities.today = {
    ...jsonData.activities.today,
    created: today,
    text: today,
  };

  return (
    <WebexDataProvider adapter={adapter}>
      <WebexActivity activityID="today" />
    </WebexDataProvider>
  );
});

stories.add('created yesterday', () => {
  const yesterday = subDays(new Date(), 1).toString();

  jsonData.activities.yesterday = {
    ...jsonData.activities.yesterday,
    created: yesterday,
    text: yesterday,
  };

  return (
    <WebexDataProvider adapter={adapter}>
      <WebexActivity activityID="yesterday" />
    </WebexDataProvider>
  );
});

stories.add('created this week', () => {
  // if it's sunday, make it a monday, otherwise pick the day before today
  const thisWeek = getDay(new Date()) === 0
    ? addDays(new Date(), 1)
    : subDays(new Date(), 2).toString();

  jsonData.activities.sameWeek = {
    ...jsonData.activities.sameWeek,
    created: thisWeek,
    text: thisWeek,
  };

  return (
    <WebexDataProvider adapter={adapter}>
      <WebexActivity activityID="sameWeek" />
    </WebexDataProvider>
  );
});

stories.add('created over a week ago', () => {
  const oldDate = subDays(new Date(), 7).toString();

  jsonData.activities.old = {
    ...jsonData.activities.old,
    created: oldDate,
    text: oldDate,
  };

  return (
    <WebexDataProvider adapter={adapter}>
      <WebexActivity activityID="old" />
    </WebexDataProvider>
  );
});
