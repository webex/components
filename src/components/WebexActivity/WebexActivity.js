import React from 'react';
import PropTypes from 'prop-types';

import {useActivity} from '../hooks';

import ActivityHeader from './ActivityHeader';
import './WebexActivity.scss';

/**
 * WebexActivity component displays activity content.
 *
 * @param {object} props
 * @returns {object} JSX of the component
 */
export default function WebexActivity({activityID}) {
  const {ID, text, created, displayHeader, personID} = useActivity(activityID);
  const header = displayHeader ? <ActivityHeader personID={personID} timestamp={created} /> : null;

  return (
    <div className="activity" key={ID}>
      {header}
      <div className="activity-content">{text}</div>
    </div>
  );
}

WebexActivity.propTypes = {
  activityID: PropTypes.string.isRequired,
};
