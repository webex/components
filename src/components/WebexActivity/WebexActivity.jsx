import React from 'react';
import PropTypes from 'prop-types';

import webexComponentClasses from '../helpers';
import {useActivity} from '../hooks';

import ActivityHeader from './ActivityHeader';
import WebexAdaptiveCard from '../WebexAdaptiveCard/WebexAdaptiveCard';

/**
 * WebexActivity component displays activity content.
 *
 * @param {object} props  Data passed to the component
 * @param {string} props.activityID  ID of the activity
 * @param {string} props.className  Custom CSS class to apply
 * @param {object} props.style  Custom style to apply
 * @returns {object} JSX of the component
 */
export default function WebexActivity({activityID, className, style}) {
  const {
    ID,
    created,
    displayHeader,
    personID,
    text,
    card,
  } = useActivity(activityID);
  const [cssClasses, sc] = webexComponentClasses('activity', className);

  return (
    <div className={cssClasses} key={ID} style={style}>
      {displayHeader && <ActivityHeader personID={personID} timestamp={created} />}
      <div className={sc('content')}>
        {text}
        {card && <WebexAdaptiveCard activityID={ID} />}
      </div>
    </div>
  );
}

WebexActivity.propTypes = {
  activityID: PropTypes.string.isRequired,
  className: PropTypes.string,
  style: PropTypes.shape(),
};

WebexActivity.defaultProps = {
  className: '',
  style: undefined,
};
