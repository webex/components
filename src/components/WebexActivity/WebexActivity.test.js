import React from 'react';

import WebexActivity from './WebexActivity';

jest.mock('../hooks/useActivity');

describe('Webex Activity', () => {
  describe('component snapshot', () => {
    test('matches snapshot with "default" text', () => {
      expect(shallow(<WebexActivity activityID="default" />)).toMatchSnapshot();
    });

    test('matches snapshot with an activity without header', () => {
      expect(shallow(<WebexActivity activityID="no-header" />)).toMatchSnapshot();
    });

    test('matches snapshot with "long" text', () => {
      expect(shallow(<WebexActivity activityID="long" />)).toMatchSnapshot();
    });

    test('matches snapshot with a text from today', () => {
      expect(shallow(<WebexActivity activityID="today" />)).toMatchSnapshot();
    });

    test('matches snapshot with a text from yesterday', () => {
      expect(shallow(<WebexActivity activityID="yesterday" />)).toMatchSnapshot();
    });

    test('matches snapshot with a text from the same week', () => {
      expect(shallow(<WebexActivity activityID="sameWeek" />)).toMatchSnapshot();
    });

    test('matches snapshot with a text over a week ago', () => {
      expect(shallow(<WebexActivity activityID="old" />)).toMatchSnapshot();
    });
  });
});
