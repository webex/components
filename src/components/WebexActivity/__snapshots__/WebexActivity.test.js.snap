// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webex Activity component snapshot matches snapshot with "default" text 1`] = `
<div
  className="activity"
  key="default"
>
  <ActivityHeader
    personID="default"
    timestamp="2019-08-20T21:00:07.047Z"
  />
  <div
    className="activity-content"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  </div>
</div>
`;

exports[`Webex Activity component snapshot matches snapshot with "long" text 1`] = `
<div
  className="activity"
  key="long"
>
  <ActivityHeader
    personID="default"
    timestamp="2019-08-20T21:00:07.047Z"
  />
  <div
    className="activity-content"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </div>
</div>
`;

exports[`Webex Activity component snapshot matches snapshot with a text from the same week 1`] = `
<div
  className="activity"
  key="sameWeek"
>
  <ActivityHeader
    personID="default"
    timestamp="sameWeek"
  />
  <div
    className="activity-content"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  </div>
</div>
`;

exports[`Webex Activity component snapshot matches snapshot with a text from today 1`] = `
<div
  className="activity"
  key="today"
>
  <ActivityHeader
    personID="default"
    timestamp="today"
  />
  <div
    className="activity-content"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  </div>
</div>
`;

exports[`Webex Activity component snapshot matches snapshot with a text from yesterday 1`] = `
<div
  className="activity"
  key="yesterday"
>
  <ActivityHeader
    personID="default"
    timestamp="yesterday"
  />
  <div
    className="activity-content"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  </div>
</div>
`;

exports[`Webex Activity component snapshot matches snapshot with a text over a week ago 1`] = `
<div
  className="activity"
  key="old"
>
  <ActivityHeader
    personID="default"
    timestamp="old"
  />
  <div
    className="activity-content"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  </div>
</div>
`;

exports[`Webex Activity component snapshot matches snapshot with an activity without header 1`] = `
<div
  className="activity"
  key="no-header"
>
  <div
    className="activity-content"
  >
    no header text
  </div>
</div>
`;
